core/common/src/main/scala/org/typelevel/otel4s/context/propagation/TextMapPropagator.scala

```diff
@@ -0,0 +1,169 @@
+  def fields: Iterable[String]
+
+  /** Extracts key-value pairs from the given `carrier` and adds them to the
+    * given context.
+    *
+    * @param ctx
+    *   the context object to which the key-value pairs are added
+    *
+    * @param carrier
+    *   holds propagation fields
+    *
+    * @tparam A
+    *   the type of the carrier object
+    *
+    * @return
+    *   the new context with stored key-value pairs
+    */
+  def extract[A: TextMapGetter](ctx: Ctx, carrier: A): Ctx
+
+  /** Injects data from the context into a copy of the given immutable `carrier`
+    * for downstream consumers, for example as HTTP headers.
+    *
+    * This method is an extension to the OpenTelemetry specification to support
+    * immutable carrier types.
+    *
+    * @param ctx
+    *   the context containing the value to be injected
+    *
+    * @param carrier
+    *   holds propagation fields
+    *
+    * @tparam A
+    *   the type of the carrier
+    *
+    * @return
+    *   a copy of the carrier, with new fields injected
+    */
+  def inject[A: TextMapUpdater](ctx: Ctx, carrier: A): A
+}
+
+object TextMapPropagator {
+
+  /** Creates a [[TextMapPropagator]] which delegates injection and extraction
+    * to the provided propagators.
+    *
+    * @example
+    *   {{{
+    * val w3cPropagator: TextMapPropagator[Context] = ???
+    * val httpTracePropagator: TextMapPropagator[Context] = ???
+    * val textMapPropagator = TextMapPropagator.of(w3cPropagator, httpTracePropagator)
+    *   }}}
+    *
+    * @param propagators
+    *   the propagators to use for injection and extraction
+    *
+    * @tparam Ctx
+    *   the context to use to extract or inject data
+    */
+  def of[Ctx](
+      propagators: TextMapPropagator[Ctx]*
+  ): TextMapPropagator[Ctx] =
+    propagators.combineAll
+
+  /** Creates a no-op implementation of the [[TextMapPropagator]].
+    *
+    * All propagation operations are no-op.
+    */
+  def noop[Ctx]: TextMapPropagator[Ctx] =
+    new Noop
+
+  implicit def textMapPropagatorMonoid[Ctx]: Monoid[TextMapPropagator[Ctx]] =
+    new Monoid[TextMapPropagator[Ctx]] {
+      val empty: TextMapPropagator[Ctx] =
+        noop[Ctx]
+
+      def combine(
+          x: TextMapPropagator[Ctx],
+          y: TextMapPropagator[Ctx]
+      ): TextMapPropagator[Ctx] =
+        (x, y) match {
+          case (that, _: Noop[Ctx]) =>
+            that
+          case (_: Noop[Ctx], other) =>
+            other
+          case (that: Multi[Ctx], other: Multi[Ctx]) =>
+            multi(that.propagators ++ other.propagators)
+          case (that: Multi[Ctx], other) =>
+            multi(that.propagators :+ other)
+          case (that, other: Multi[Ctx]) =>
+            multi(that +: other.propagators)
+          case (that, other) =>
+            multi(List(that, other))
+        }
+
+      private def multi(propagators: List[TextMapPropagator[Ctx]]): Multi[Ctx] =
+        Multi(propagators, propagators.flatMap(_.fields))
```

Comment:
should the fields be `.distinct`?

Reply:
Good catch. You are correct. I added the `.distinct` call and a test.